pipewire.md: add a page on setting up pipewire (#540)
void-linux/void-docs: src/config/media/pipewire.md

# PipeWire

To use PipeWire, install the `pipewire` package.

PipeWire should be started as a user. Use the autostarting mechanism of your
desktop environment. If using [startx](../graphical-session/xorg.md#startx),
add the following line in your `.xinitrc` file:

```
pipewire &
```

## PulseAudio replacement

> PipeWire can start a PulseAudio server. To enable it, uncomment the following

Reviewer: I would now suggest to launch `pipewire-pulse` in the same place where you launch `pipewire`, instead of from the config. Keeping the config unmodified allows xbps to automatically replace it with the updated versions.

Author: That's a great suggestion.
I've tried this on Sway, by executing both `pipewire` and `pipewire-pulse` and it works fine.

line in `/etc/pipewire/pipewire.conf` and restart the `pipewire` daemon:

```
exec /usr/bin/pipewire-pulse
```

If PulseAudio service is enabled,
[disable](../services/index.md#disabling-services) it and reboot.

To check if the replacement is working correctly, use the following command:

```
$ pactl info

...
Server Name: PulseAudio (on PipeWire 0.3.18)
...
```

## ALSA integration

> Install `alsa-pipewire`, then enable the pipewire alsa device and make it

Reviewer 2: ```suggestion
Install `alsa-pipewire`, then enable the PipeWire ALSA device and make it
```

default:

```
# mkdir -p /etc/alsa/conf.d
# ln -s /usr/share/alsa/alsa.conf.d/50-pipewire.conf /etc/alsa/conf.d
# ln -s /usr/share/alsa/alsa.conf.d/99-pipewire-default.conf /etc/alsa/conf.d
```

## JACK replacement

Install `libjack-pipewire` and use `pw-jack` to launch JACK clients:

```
$ pw-jack <application>

> ```

Reviewer: here you could mention how to set up the automatic libjack override via `/etc/ld.so.conf.d` (glibc) or `/etc/ld-musl-$ARCH.path` (musl)

Author: I'm not sure what exactly this suggestion should be -- never worked with JACK.
I can see some notes on this in your gist, but I'm not sure what the actual steps should be. Can you give an example?

Reviewer 3: @kkga I these steps are available [on Reddit](https://www.reddit.com/r/voidlinux/comments/mjzlry/replacing_jack_with_pipewire/gteamw8?utm_source=share&utm_medium=web2x&context=3).

For glibc:
1. Create `/etc/ld.so.conf.d/pipewire-jack.conf` with content `/usr/lib/pipewire-0.3/jack`
2. `sudo ldconfig`

Worked for me.

Author: Added a note for glibc.
Can you post the exact steps needed for musl? Don't want to mess this up, since I have no way of testing.

Reviewer: I have just tried it for musl and it's too easy to fuck up, so I don't feel like recommending it. Basically musl will not fall back to `/lib` unless explicitly told so, so creating that file is dangerous.

Reviewer 4: What is the problem exactly for musl? I remember it working effortlessly, although right now I am on glibc and cannot test.
I would prefer for the instructions to be complete with respect to musl as well.

Reviewer 2: Writing a path to musl's config file and forgetting to also include `/lib` means your system won't find any libraries. And now you can't even turn off your computer because `sudo` doesn't work.

# PipeWire

To use PipeWire, install the `pipewire` package.

[pipewire(1)](https://man.voidlinux.org/pipewire) should be started as a user.
Use the autostarting mechanism of your desktop environment or
[startx](../graphical-session/xorg.md#startx).

## PulseAudio replacement

PipeWire can start a PulseAudio server. While it's possible to modify
`/etc/pipewire/pipewire.conf` for auto-starting the PulseAudio server, it's
recommended to start it from the same place where you start PipeWire by
executing `pipewire-pulse`. This keeps PipeWire's configuration file unmodified
for smoother future upgrades.

After setting up `pipewire-pulse`, make sure that the PulseAudio service is
[disabled](../services/index.md#disabling-services) and reboot.

To check if the replacement is working correctly, use

> [pactl(1)](https://man.voidlinux.org/pactl):

Reviewer 5: What do you think about mentioning the package which contains pactl here? It was recently split from pulseaudio to pulseaudio-utils.

# PipeWire

To use PipeWire, install the `pipewire` package.

> [pipewire(1)](https://man.voidlinux.org/pipewire) should be started as a user.

Reviewer 2: ```suggestion
[pipewire(1)](https://man.voidlinux.org/pipewire.1) should be started as a user.
```

Use the autostarting mechanism of your desktop environment or
[startx](../graphical-session/xorg.md#startx).

## PulseAudio replacement

PipeWire can start a PulseAudio server. While it's possible to modify

> `/etc/pipewire/pipewire.conf` for auto-starting the PulseAudio server, it's

Reviewer 2: ```suggestion
`[pipewire.conf(5)](link)` for auto-starting the PulseAudio server, it's
```

recommended to start it from the same place where you start PipeWire by
executing `pipewire-pulse`. This keeps PipeWire's configuration file unmodified
for smoother future upgrades.

After setting up `pipewire-pulse`, make sure that the PulseAudio service is

> [disabled](../services/index.md#disabling-services) and reboot.

Reviewer 2: Reboot sounds like the wrong advice. I think this should be worded to "make sure no other PulseAudio server instance is running".

To check if the replacement is working correctly, use

> [pactl(1)](https://man.voidlinux.org/pactl) (provided by the `pulseaudio-utils`

Reviewer 2: ```suggestion
[pactl(1)](https://man.voidlinux.org/pactl.1) (provided by the `pulseaudio-utils`
```

package):

```
$ pactl info

> ...
> Server Name: PulseAudio (on PipeWire 0.3.18)
> ...

Reviewer 2: ```suggestion
[...]
Server Name: PulseAudio (on PipeWire 0.3.18)
[...]
```

```

> ##### Bluetooth audio

Reviewer 2: I think this should be `##`?

For bluetooth audio to work, install the `libspa-bluetooth` package.

## ALSA integration

Install `alsa-pipewire`, then enable the PipeWire ALSA device and make it
default:

```
# mkdir -p /etc/alsa/conf.d
# ln -s /usr/share/alsa/alsa.conf.d/50-pipewire.conf /etc/alsa/conf.d
# ln -s /usr/share/alsa/alsa.conf.d/99-pipewire-default.conf /etc/alsa/conf.d
```

## JACK replacement

Install `libjack-pipewire`.

> Use [pw-jack(1)](https://man.voidlinux.org/pw-jack) to launch JACK clients

Reviewer 2: ```suggestion
Use [pw-jack(1)](https://man.voidlinux.org/pw-jack.1) to launch JACK clients
```

manually:

```
$ pw-jack <application>
```

> Alternatively, override the library provided by `libjack`. The following
> approach will work on `glibc`:

Reviewer 2: ```suggestion
Alternatively, override the library provided by `libjack`. The following
approach will work on glibc-based systems:
```

> approach will work on `glibc`:

Reviewer 2: Maybe link `ld.so(8)`?

Reviewer 2: I prefer helping at least a bit. You can still copy from it, but you can also read to try and understand what is being done :)

Author: Right, i've added a link.

# PipeWire

To use PipeWire, install the `pipewire` package.

[pipewire(1)](https://man.voidlinux.org/pipewire.1) should be started as a user.
Use the autostarting mechanism of your desktop environment or
[startx](../graphical-session/xorg.md#startx).

## PulseAudio replacement

PipeWire can start a PulseAudio server. While it's possible to modify
[pipewire.conf(5)](https://man.voidlinux.org/pipewire.conf.5) for auto-starting
the PulseAudio server, it's recommended to start it from the same place where

> you start PipeWire by executing `pipewire-pulse`. This keeps PipeWire's

Reviewer 5: "PipeWire's configuration file" sounds a bit weird to me, what about "the PipeWire configuration file"? Keep in mind that I am not a native speaker, so I might be wrong.

Author: Correct, I'll update this

configuration file unmodified for smoother future upgrades.

After setting up `pipewire-pulse`, make sure that the PulseAudio service is
[disabled](../services/index.md#disabling-services) and make sure no other

> PulseAudio server instance is running.

Reviewer 5: I think the word "After" is not exactly right here, people should check if the pulseaudio server is not running **before** "setting up" pipewire-pulse.

Also, "setting up" is not exactly right, I think it should be something like "starting" or "running".

Author: Sounds good

To check if the replacement is working correctly, use
[pactl(1)](https://man.voidlinux.org/pactl.1) (provided by the
`pulseaudio-utils` package):

```
$ pactl info

[...]
Server Name: PulseAudio (on PipeWire 0.3.18)
[...]
```

## Bluetooth audio

For bluetooth audio to work, install the `libspa-bluetooth` package.

## ALSA integration

> Install `alsa-pipewire`, then enable the PipeWire ALSA device and make it
> default:

Reviewer 2: ```suggestion
Install `alsa-pipewire`, then enable the PipeWire ALSA device and make it
the default:
```

# PipeWire

To use PipeWire, install the `pipewire` package.

PipeWire should be started as a user. Use the autostarting mechanism of your

> desktop environment.

Reviewer 2: Linking `[pipewire(1)](link)` somewhere around here would be nice too.

If using [startx](../graphical-session/xorg.md#startx), add the following line
in your `.xinitrc` file:

```
pipewire &

> ```

Reviewer 2: This should be a note in the startx section about how to start arbitrary programs together with your X session.

Author: Thanks, I've added a note on this in the startx section. Let me know if it looks good to you.

If using [Sway](https://swaywm.org/), start it from the configuration file:

```
exec pipewire

> ```

Reviewer 2: This shouldn't be here, IMO.

## PulseAudio replacement

PipeWire can start a PulseAudio server. While it's possible to modify
`/etc/pipewire/pipewire.conf` for auto-starting the PulseAudio server, it's
recommended to start it from the same place where you start PipeWire by
executing `pipewire-pulse`. This keeps PipeWire's configuration file unmodified
for smoother future upgrades.

#### Example for `.xinitrc`

```
pipewire &
pipewire-pulse &
```

#### Example for Sway

```
exec pipewire
exec pipewire-pulse

> ```

Reviewer 2: This example can be inferred from above and isn't needed.

After setting up `pipewire-pulse`, make sure that the PulseAudio service is

> [disabled](../services/index.md#disabling-services) and reboot.

Reviewer 2: More important is probably to disable PulseAudio auto activation (I can add a section on that), because most people shouldn't be using the service already.

Author: What do you mean by disabling auto activation?
I think pulseaudio should not start automatically if the pipewire-pulse server is running.

Reviewer 5: Pulseaudio can start automatically before pipewire-pulse is ran which creates a conflict between them.

> To check if the replacement is working correctly, use the following command:

Reviewer 2: ```suggestion
To check if the replacement is working correctly, use [pactl(1)](link):
```

```
$ pactl info

...
Server Name: PulseAudio (on PipeWire 0.3.18)
...
```

##### Note on Bluetooth audio

If having troubles with Bluetooth audio, make sure that `libspa-bluetooth`

> package is installed.

Reviewer 2: The package is required for bluetooth at all, so I don't think this is the correct language.

And title should be just `Bluetooth audio`

Author: It was required for bluetooth audio to work in my case. But let me know if we should remove this.

Reviewer 5: What @ericonr is saying is that "If having troubles" is not the correct language when libspa-bluetooth is required for bluetooth audio to work.

Author: Oh I misread that, I thought @ericonr said "is _not_ required for bluetooth at all" :)

Author: updated

## ALSA integration

Install `alsa-pipewire`, then enable the PipeWire ALSA device and make it
default:

```
# mkdir -p /etc/alsa/conf.d
# ln -s /usr/share/alsa/alsa.conf.d/50-pipewire.conf /etc/alsa/conf.d
# ln -s /usr/share/alsa/alsa.conf.d/99-pipewire-default.conf /etc/alsa/conf.d

> ```

Reviewer 2: I think this section could have a link to `alsa` docs on its configuration directories, if you can find something.

## JACK replacement

Install `libjack-pipewire`.

> Use `pw-jack` to launch JACK clients manually:

Reviewer 2: ```suggestion
Use [pw-jack(1)](link to manpage) to launch JACK clients manually:
```

```
$ pw-jack <application>
```

> Alternatively, override `libjack`:

Reviewer 2: ```suggestion
Alternatively, override the library provided by `libjack`:
```

```
# echo "/usr/lib/pipewire-0.3/jack" > /etc/ld.so.conf.d/pipewire-jack.conf
# ldconfig

> ```

Reviewer 2: You should mention this is glibc specific.
